eth/transactionManager_test.go

```diff
@@ -237,6 +237,7 @@ func TestTransactionManager_Replace(t *testing.T) {
 	logsAfter = glog.Stats.Info.Lines()
 	assert.Nil(err)
 	expTx := types.NewTransaction(1, *stubTx.To(), stubTx.Value(), 100000, calcReplacementGasPrice(stubTx), stubTx.Data())
+	assert.Equal(tx.Hash(), expTx.Hash())
```

Comment:
Fun discovery that led to the addition of this assertion as well as the assertion on L253:

The go-ethereum `Transaction` type maintains a cache for the hash of the tx and the `Hash()` method will only compute the hash of the tx if the [cache is not populated](https://github.com/ethereum/go-ethereum/blob/28ba686cbfa6a46a9b4f03b620460867d502f34e/core/types/transaction.go#L369). The internal [hash field](https://github.com/ethereum/go-ethereum/blob/28ba686cbfa6a46a9b4f03b620460867d502f34e/core/types/transaction.go#L56) of the struct is the zero value until `Hash()` is called for the first time. 

Prior to this PR, `newSignedTx.Hash()` was not called in `replace()` when an error was not returned from `SubmitTransaction()`. But, now `newSignedTx.Hash()` is called in `replace()` when an error is not returned from `SubmitTransaction()` because we call `newSignedTx.Hash()` in the log statement on L188. As a result, without these assertions that call the `Hash()` method on both `tx` and `expTx`, one of the structs would have the zero value for the internal hash field and the assertion comparing `tx` and `expTx` would fail. By first calling `Hash()` on both `tx` and `expTx` we ensure that the internal hash field of both structs is populated.

Reply:
Oh nice find, interesting behaviour.